flaml/scheduler/online_scheduler.py

```diff
@@ -0,0 +1,163 @@
+    """  Keep the top performed learners running
+
+    Methods:
+        on_trial_result(trial_runner, trial, result)
+            Report result and return a decision on the trial's status
+        choose_trial_to_run(trial_runner)
+            Decide which trial to run next
+    """
+    def __init__(self,
+                 max_lease: Optional[float] = np.inf,
+                 increase_factor: Optional[float] = 2,
+                 **kwargs
+                 ):
+        '''
+        Args:
+            max_lease: float = np.inf
+            increase_factor: float = 2
+        '''
+        super().__init__(max_lease, increase_factor)
+        self._keep_champion = kwargs.get('keep_champion', True)
+        self._keep_challenger_metric = kwargs.get('keep_challenger_metric', 'ucb')
+        self._keep_challenger_ratio = kwargs.get('keep_challenger_ratio', 'tophalf')
+        self._pause_old_froniter = kwargs.get('pause_old_froniter', False)
+        logger.info('Using chacha scheduler with config %s', kwargs)
+
+    def on_trial_result(self, trial_runner, trial: Trial, result: Dict):
+        """Report result and return a decision on the trial's status
+
+        Make a decision according to: SuccessiveDoubling + champion check + performance check
+        """
+        # Doubling scheduler makes a decision
+        decision = super().on_trial_result(trial_runner, trial, result)
+        # ***********Check whether the trial has been paused since a new champion is promoted**
+        ## check champion froniter if the trial is not champion froniter, it means it
+        # has not been paused since the new champion is promoted. If so, we need to
+        # tentative pause it such that new trials can possiblly be taken into consideration
+        # NOTE: this part may need to be changed. We need to do this because we only add trials
+        # into the OnlineTrialRunner when there are avaialbe slots. Maybe we need to consider
+        # adding max_running_trial number of trials once a new champion is promoted.
+        if self._pause_old_froniter and not trial.is_checked_under_current_champion:
+            if decision != TrialScheduler.STOP:
+                decision = TrialScheduler.PAUSE
+                trial.set_checked_under_current_champion(True)
+                logger.info('Tentitively set trial as paused')
+
+        # ****************Keep the champion always running******************
+        if self._keep_champion and trial.trial_id == trial_runner.champion_trial.trial_id and \
+           decision == TrialScheduler.PAUSE:
+            return TrialScheduler.CONTINUE
+
+        # ****************Keep the trials with top performance always running******************
+        if self._keep_challenger_ratio is not None:
+            if decision == TrialScheduler.PAUSE:
+                logger.debug('champion, %s', trial_runner.champion_trial.trial_id)
+                top_trials = trial_runner.get_top_running_trials(self._keep_challenger_ratio,
```

Comment:
This way of checking may be inefficient when the # trials is large. Add a comment about it.